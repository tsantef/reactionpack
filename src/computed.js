import _ from 'lodash';

function getSelectorValues(state, selectors) {
	return _.map(selectors, (selector) => selector(state));
}

export function createComputed(args) {
	if (_.size(args) < 2) {
		throw new Error('Missing both a selector and compute function');
	}
	const selectors = _.initial(args);
	const compute = _.last(args);
	let resultCache;
	let valueCache;
	return function(state) {
		const selectorValues = getSelectorValues(state, selectors);
		if (valueCache && _.every(selectorValues, (value, index) => (value === valueCache[index]))) {
			return resultCache;
		}
		valueCache = selectorValues;
		resultCache = compute.apply(null, selectorValues);
		return resultCache;
	};
}

export function buildComputeds(computeds) {
	return _.mapValues(computeds, (value, key) => {
		if (_.isArray(value)) {
			return createComputed(value);
		} else {
			return buildComputeds(value);
		}
	});
}

export function generateComputeds(computeds, props, nameSpace) {
	const _props = nameSpace ? _.get(props, nameSpace, {}) : props;
	return _.mapValues(computeds, (computed, key) => {
		if (_.isFunction(computed)) {
			return computed(_props);
		} else {
			return generateComputeds(computed, _props, key);
		}
	});
}
