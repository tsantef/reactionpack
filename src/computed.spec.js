import { createComputed } from './computed';

describe('Computed', () => {

	describe('#createComputed', () => {

		it('should compute new value', () => {
			let callCount = 0;

			const testComputed = createComputed([
				(state) => state.value,
				(inputText) => {
					callCount = callCount + 1;
					return callCount;
				},
			]);

			expect(testComputed({ value: 'test1' })).toEqual(1);
			expect(testComputed({ value: 'test2' })).toEqual(2);
		});

		it('should not recompute same value', () => {
			let callCount = 0;

			const testComputed = createComputed([
				(state) => state.value,
				(inputText) => {
					callCount = callCount + 1;
					return callCount;
				},
			]);

			expect(testComputed({ value: 'test' })).toEqual(1);
			expect(testComputed({ value: 'test' })).toEqual(1);
		});

		it('should throw error if missing a selector or computed function', () => {
			expect(() => {
				createComputed();
			}).toThrowErrorMatchingSnapshot();
			expect(() => {
				createComputed([]);
			}).toThrowErrorMatchingSnapshot();
			expect(() => {
				createComputed([
					(state) => {},
				]);
			}).toThrowErrorMatchingSnapshot();
		});

		it('should work with multiple selectors', () => {
			const testComputed = createComputed([
				(state) => state.value1,
				(state) => state.value2,
				(value1, value2) => {
					return value1 + value2;
				},
			]);

			expect(testComputed({ value1: 1, value2: 2 })).toEqual(3);
		});

		it('should compute after not recomputing', () => {

			let counter = 2;

			const state1 = {
				value: 2,
			};

			const state1dup = {
				value: 2,
			};

			const state2 = {
				value: 5,
			};

			const computed = [
				(state) => state.value,
				(value) => {
					return value * counter++;
				},
			];

			const testComputed = createComputed(computed);

			expect(testComputed(state1)).toEqual(4);
			expect(testComputed(state1dup)).toEqual(4);
			expect(testComputed(state2)).toEqual(15);

		});

	});

});
