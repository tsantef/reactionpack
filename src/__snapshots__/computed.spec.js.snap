// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Computed #createComputed should throw error if missing a selector or computed function 1`] = `"Missing both a selector and compute function"`;

exports[`Computed #createComputed should throw error if missing a selector or computed function 2`] = `"Missing both a selector and compute function"`;

exports[`Computed #createComputed should throw error if missing a selector or computed function 3`] = `"Missing both a selector and compute function"`;
